### PR TITLE
fix(ext/node): fix vm memory usage and context initialization

### PR DESCRIPTION
**Pull request by littledivy**
Fixes https://github.com/denoland/deno/issues/22441
Fixes https://github.com/denoland/deno/issues/23913
Fixes  https://github.com/denoland/deno/issues/23852
Fixes https://github.com/denoland/deno/issues/23917

### CODE REVIEW

**Comment by littledivy** [comment]
External reference missing only on Aarch64 linux release mode. From the runner:

```
runner@vm6stvbz:~/work/deno/deno$ sudo lldb
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'lldb.embedded_interpreter'
(lldb) attach -w -n /home/runner/work/deno/deno/target/release/build/deno-ad605ac29f6acc83/build-script-build
Process 48452 stopped
* thread #1, name = 'build-script-bu', stop reason = signal SIGSTOP
    frame #0: 0x0000fffff7fd9c40 ld-linux-aarch64.so.1`___lldb_unnamed_symbol219 + 1072
ld-linux-aarch64.so.1`___lldb_unnamed_symbol219:
->  0xfffff7fd9c40 <+1072>: bti    c
    0xfffff7fd9c44 <+1076>: mov    x0, sp
    0xfffff7fd9c48 <+1080>: bl     0xfffff7fda610            ; ___lldb_unnamed_symbol234
    0xfffff7fd9c4c <+1084>: mov    x21, x0
Executable module set to "/home/runner/work/deno/deno/target/release/build/deno-ad605ac29f6acc83/build-script-build".
Architecture set to: aarch64-unknown-linux-gnu.
(lldb) c
Process 48452 resuming
Process 48452 stopped and restarted: thread 1 received signal: SIGCHLD
Process 48452 stopped and restarted: thread 1 received signal: SIGCHLD
Process 48452 stopped
* thread #1, name = 'build-script-bu', stop reason = signal SIGTRAP
    frame #0: 0x0000aaaaacd54ef0 build-script-build`v8::base::OS::Abort() + 48
build-script-build`v8::base::OS::Abort:
->  0xaaaaacd54ef0 <+48>: brk    #0
    0xaaaaacd54ef4 <+52>: hlt    #0

build-script-build`v8::base::OS::DebugBreak:
    0xaaaaacd54ef8 <+0>:  bti    c
    0xaaaaacd54efc <+4>:  brk    #0
(lldb) image lookup --address 0xaaaaaf36f874
      Address: build-script-build[0x00000000048cf874] (build-script-build.PT_LOAD[1]..text + 40089716)
      Summary: build-script-build`deno_node::ops::vm_internal::c_noop::h5fb870596cec6342
(lldb)
```

File: .github/workflows/ci.generate.ts [hunk 1]
```diff
@@ -688,6 +688,16 @@ const ci = {
...
+        // Uncomment for remote debugging
+        // {
+        //   name: "Setup tmate session",
+        //   if: [
+        //     "(matrix.job == 'test' || matrix.job == 'bench') &&",
+        //     "matrix.profile == 'release' && (matrix.use_sysroot ||",
+        //     "github.repository == 'denoland/deno')",
+        //   ].join("\n"),
+        //   uses: "mxschmitt/action-tmate@v3",
+        // },
```
Comment (bartlomieju) [inline .github/workflows/ci.generate.ts:700]: Is this left on purpose?

File: ext/node/ops/vm.rs [hunk 2]
```diff
@@ -7,6 +7,8 @@ use deno_core::v8;
 
 use super::vm_internal as i;
 
+pub use i::*;
```
Comment (bartlomieju) [inline ext/node/ops/vm.rs:10]: Do we need to expose all these symbols?

File: ext/node/lib.rs [hunk 3]
```diff
@@ -641,7 +644,64 @@ deno_core::extension!(deno_node,
   global_template_middleware = global_template_middleware,
   global_object_middleware = global_object_middleware,
   customizer = |ext: &mut deno_core::Extension| {
-    let mut external_references = Vec::with_capacity(7);
+    let mut external_references = Vec::with_capacity(14);
+
+    vm::GETTER_MAP_FN.with(|getter| {
```
Comment (bartlomieju) [inline ext/node/lib.rs:649]: We should try to remove this thread local hack - maybe the problem is now gone

File: ext/node/ops/vm_internal.rs [hunk 4]
```diff
@@ -177,17 +186,25 @@ impl ContextifyContext {
 
 pub const VM_CONTEXT_INDEX: usize = 0;
 
-fn create_v8_context<'a>(
-  scope: &mut v8::HandleScope<'a>,
+pub fn create_v8_context<'a>(
+  scope: &mut v8::HandleScope<'a, ()>,
   object_template: v8::Local<v8::ObjectTemplate>,
-  snapshot_data: Option<&'static [u8]>,
+  use_snapshot: bool,
```
Comment (bartlomieju) [inline ext/node/ops/vm_internal.rs:192]: Consider using an enum instead, it's really not obvious what these bools mean at the call site

File: ext/node/ops/vm_internal.rs [hunk 5]
```diff
@@ -196,25 +213,31 @@ fn create_v8_context<'a>(
 #[derive(Debug, Clone)]
 struct SlotContextifyGlobalTemplate(v8::Global<v8::ObjectTemplate>);
 
-fn init_global_template<'a>(
-  scope: &mut v8::HandleScope<'a>,
+pub fn init_global_template<'a>(
+  scope: &mut v8::HandleScope<'a, ()>,
+  cache: bool,
```
Comment (bartlomieju) [inline ext/node/ops/vm_internal.rs:218]: Ditto

File: ext/node/lib.rs (hunk 3 above)
Comment (bartlomieju) [reply ext/node/lib.rs:649]: But not in this PR

File: .github/workflows/ci.generate.ts (hunk 1 above)
Comment (littledivy) [reply .github/workflows/ci.generate.ts:700]: yes, ive used it quite a few times for segfaults in CIs

File: ext/node/ops/vm.rs (hunk 2 above)
Comment (littledivy) [reply ext/node/ops/vm.rs]: Fixed

**Pull request merged**